src/promql/src/planner.rs

```diff
@@ -0,0 +1,627 @@
+            .context(DataFusionPlanningSnafu)?
+            .build()
+            .context(DataFusionPlanningSnafu)?;
+        Ok(result)
+    }
+
+    /// Setup [PromPlannerContext]'s state fields.
+    fn setup_context(&mut self, table_name: &str) -> Result<()> {
+        let table = self
+            .schema_provider
+            .get_table_provider(TableReference::Bare { table: table_name })
+            .context(DataFusionPlanningSnafu)?
+            .as_any()
+            .downcast_ref::<DefaultTableSource>()
+            .context(UnknownTableSnafu)?
+            .table_provider
+            .as_any()
+            .downcast_ref::<DfTableProviderAdapter>()
+            .context(UnknownTableSnafu)?
+            .table();
+
+        // set time index column name
+        let time_index = table
+            .schema()
+            .timestamp_column()
+            .with_context(|| TimeIndexNotFoundSnafu { table: table_name })?
+            .name
+            .clone();
+        self.ctx.time_index_column = Some(time_index);
+
+        // set values column
+        let values = table
+            .table_info()
+            .meta
+            .value_column_names()
+            .cloned()
+            .collect();
+        self.ctx.value_columns = values;
+
+        Ok(())
+    }
+
+    // TODO(ruihang): insert column expr
+    fn create_function_args(&self, args: &[Box<PromExpr>]) -> Result<FunctionArgs> {
+        let mut result = FunctionArgs::default();
+
+        for arg in args {
+            match *arg.clone() {
+                PromExpr::AggregateExpr { .. }
+                | PromExpr::UnaryExpr { .. }
+                | PromExpr::BinaryExpr { .. }
+                | PromExpr::ParenExpr { .. }
+                | PromExpr::SubqueryExpr { .. }
+                | PromExpr::VectorSelector { .. }
+                | PromExpr::MatrixSelector { .. }
+                | PromExpr::Call { .. } => {
+                    if result.input.replace(*arg.clone()).is_some() {
+                        MultipleVectorSnafu { expr: *arg.clone() }.fail()?;
+                    }
+                }
+
+                PromExpr::NumberLiteral { val, .. } => {
+                    let scalar_value = ScalarValue::Float64(Some(val));
+                    result.literals.push(DfExpr::Literal(scalar_value));
+                }
+                PromExpr::StringLiteral { val, .. } => {
+                    let scalar_value = ScalarValue::Utf8(Some(val));
+                    result.literals.push(DfExpr::Literal(scalar_value));
+                }
+            }
+        }
+
+        Ok(result)
+    }
+
+    fn create_function_expr(
+        &self,
+        func: &Function,
+        mut other_input_exprs: Vec<DfExpr>,
+    ) -> Result<Vec<DfExpr>> {
+        // TODO(ruihang): check function args list
+
+        // TODO(ruihang): set this according to in-param list
+        let vector_pos = 0;
```

Comment:
Is column_pos or value_column_pos more proper? This is the index of the ctx.value_columns.